packages/json/src/browser/json-client-contribution.ts

```diff
@@ -49,19 +52,40 @@ export class JsonClientContribution extends BaseLanguageClientContribution {
     }
 
     protected updateSchemas(client: ILanguageClient): void {
-        const allConfigs = [...this.jsonSchemaStore.getJsonSchemaConfigurations()];
-        const config = this.preferences['json.schemas'];
-        if (config instanceof Array) {
-            allConfigs.push(...config);
-        }
         const registry: { [pattern: string]: string[] } = {};
-        for (const s of allConfigs) {
-            if (s.fileMatch) {
-                for (let fileMatch of s.fileMatch) {
-                    if (fileMatch.charAt(0) !== '/' && !fileMatch.match(/\w+:/)) {
+
+        const schemaStoreConfigs = [...this.jsonSchemaStore.getJsonSchemaConfigurations()];
+        for (const schemaConfig of schemaStoreConfigs) {
+            if (schemaConfig.fileMatch) {
+                for (let fileMatch of schemaConfig.fileMatch) {
+                    if (!fileMatch.startsWith('/') && !fileMatch.match(/\w+:/)) {
                         fileMatch = '/' + fileMatch;
                     }
-                    registry[fileMatch] = [s.url];
+                    registry[fileMatch] = [schemaConfig.url];
+                }
+            }
+        }
+
+        const preferenceConfigs = this.preferences['json.schemas'];
+        for (const schemaConfig of preferenceConfigs) {
+            if (schemaConfig.fileMatch) {
+                for (let fileMatch of schemaConfig.fileMatch) {
+                    if (!fileMatch.startsWith('/') && !fileMatch.match(/\w+:/)) {
+                        fileMatch = '/' + fileMatch;
+                    }
+
+                    const fileUri = new URI(schemaConfig.url);
+                    if (fileUri.scheme === 'file') {
+                        const filePath = fileUri.path.toString();
+                        const workspaceRootPath = this.workspace.rootPath;
```

Comment:
It does not handle multi root case. We should fetch `json.schemas` scoped by resource URI from each root and then apply such schemas only in the scope of a corresponding root.